### ISSUE TITLE
Cannot verify Azure oauth2 token signature of a valid token.

### ISSUE THREAD

**Issue by konradmalik** [lestrrat-go/jwx#395]
**Describe the bug**
Cannot verify Azure oauth2 token signature, while token is definitely good and valid - allows me to access Microsoft Graph.

The error that shows up while I want to validate the token:
```
failed to parse payload: failed to verify jws signature: failed to verify message: crypto/rsa: verification error
```

Pretty sure I must be doing something wrong, but I cannot identify what it is.

go version:
```
go version go1.16.4 darwin/amd64
```

**To Reproduce / Expected behavior**
I'm attaching a full code that uses an existing app registration, assigns a device code, created a cached refresh token, and requests a new token:

main.go
```go
package main

import (
	"context"
	"fmt"
	"io"
	"log"
	"net/http"
	"strings"
	"time"

	"github.com/AzureAD/microsoft-authentication-library-for-go/apps/public"
	"github.com/lestrrat-go/jwx/jwk"
	"github.com/lestrrat-go/jwx/jwt"
)

var cacheAccessor = &TokenCache{"serialized_cache.json"}

func acquireTokenDeviceCode() string {
	config := CreateConfig("config.json")

	app, err := public.New(config.ClientID, public.WithCache(cacheAccessor), public.WithAuthority(config.Authority))
	if err != nil {
		panic(err)
	}

	// look in the cache to see if the account to use has been cached
	var userAccount public.Account
	accounts := app.Accounts()
	for _, account := range accounts {
		if strings.EqualFold(account.PreferredUsername, config.Username) {
			userAccount = account
		}
	}
	// found a cached account, now see if an applicable token has been cached
	// NOTE: this API conflates error states, i.e. err is non-nil if an applicable token isn't
	//       cached or if something goes wrong (making the HTTP request, unmarshalling, etc).
	authResult, err := app.AcquireTokenSilent(context.Background(), config.Scopes, public.WithSilentAccount(userAccount))
	if err != nil {
		// either there was no cached account/token or the call to AcquireTokenSilent() failed
		// make a new request to AAD
		ctx, cancel := context.WithTimeout(context.Background(), 100*time.Second)
		defer cancel()
		devCode, err := app.AcquireTokenByDeviceCode(ctx, config.Scopes)
		if err != nil {
			panic(err)
		}
		fmt.Printf("Device Code is: %s\n", devCode.Result.Message)
		result, err := devCode.AuthenticationResult(ctx)
		if err != nil {
			panic(fmt.Sprintf("got error while waiting for user to input the device code: %s", err))
		}
		fmt.Println("Access token is " + result.AccessToken)
		fmt.Printf("Granted scopes are %v\n", authResult.GrantedScopes)
		return authResult.AccessToken
	}
	fmt.Println("Access token is " + authResult.AccessToken)
	fmt.Printf("Granted scopes are %v\n", authResult.GrantedScopes)
	return authResult.AccessToken
}

func callMicrosoftGraph(token string) {
	url := "https://graph.microsoft.com/v1.0/me"

	// Create a Bearer string by appending string access token
	var bearer = "Bearer " + token

	// Create a new request using http
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		panic(err)
	}

	// add authorization header to the req
	req.Header.Add("Authorization", bearer)

	// Send req using http Client
	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		log.Println("Error on response.\n[ERROR] -", err)
	}
	defer resp.Body.Close()

	// read into string
	bytes, err := io.ReadAll(resp.Body)
	if err != nil {
		log.Println("Error while reading the response bytes:", err)
	}
	fmt.Println(string(bytes))
}

func getJWTkeyset() (jwk.Set, error) {
	// TODO get this url from https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration
	const jwksURL = `https://login.microsoftonline.com/common/discovery/v2.0/keys`

	ctx := context.TODO()
	ar := jwk.NewAutoRefresh(ctx)

	// Tell *jwk.AutoRefresh that we only want to refresh this JWKS
	// when it needs to (based on Cache-Control or Expires header from
	// the HTTP response). If the calculated minimum refresh interval is less
	// than 15 minutes, don't go refreshing any earlier than 15 minutes.
	ar.Configure(jwksURL, jwk.WithMinRefreshInterval(15*time.Minute))

	// Refresh the JWKS once before getting into the main loop.
	// This allows you to check if the JWKS is available before we start
	// a long-running program
	_, err := ar.Refresh(ctx, jwksURL)
	if err != nil {
		fmt.Printf("failed to refresh JWKS: %s\n", err)
		return nil, err
	}

	keyset, err := ar.Fetch(ctx, jwksURL)
	if err != nil {
		fmt.Printf("failed to fetch JWKS: %s\n", err)
		return nil, err
	}
	return keyset, nil
}

func validateToken(tokenString string) bool {
	log.Println("Validating token")

	keyset, err := getJWTkeyset()
	if err != nil {
		panic(err)
	}

	// Actual verification:
	// FINALLY. This is how you Parse and verify the payload.
	// Key IDs are automatically matched.
	// There was a lot of code above, but as a consumer, below is really all you need
	// to write in your code
	token, err := jwt.ParseString(
		tokenString,
		// Tell the parser that you want to use this keyset
		// this will validate signature
		// TODO this does not work! why?
		jwt.WithKeySet(keyset),
		// validate expiration etc.
		jwt.WithValidate(true),
	)
	_ = keyset
	if err != nil {
		fmt.Printf("failed to parse payload: %s\n", err)
		return false
	}
	fmt.Printf("parsed token %v\n", token)
	tid, _ := token.Get("tid")
	fmt.Printf("parsed tid %v\n", tid)
	return true
}

func main() {
	token := acquireTokenDeviceCode()
	valid := validateToken(token)
	log.Printf("valid token: %t", valid)
	callMicrosoftGraph(token)
}
```

cache_accessor.go:
```go
package main

import (
	"io/ioutil"
	"log"
	"os"

	"github.com/AzureAD/microsoft-authentication-library-for-go/apps/cache"
)

type TokenCache struct {
	file string
}

func (t *TokenCache) Replace(cache cache.Unmarshaler, key string) {
	jsonFile, err := os.Open(t.file)
	if err != nil {
		log.Println(err)
	}
	defer jsonFile.Close()
	data, err := ioutil.ReadAll(jsonFile)
	if err != nil {
		log.Println(err)
	}
	err = cache.Unmarshal(data)
	if err != nil {
		log.Println(err)
	}
}

func (t *TokenCache) Export(cache cache.Marshaler, key string) {
	data, err := cache.Marshal()
	if err != nil {
		log.Println(err)
	}
	err = ioutil.WriteFile(t.file, data, 0600)
	if err != nil {
		log.Println(err)
	}
}
```

utils.go
```go
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

package main

import (
	"encoding/json"
	"io/ioutil"
	"log"
	"os"
)

// Config represents the config.json required to run the samples
type Config struct {
	ClientID  string   `json:"client_id"`
	Authority string   `json:"authority"`
	Scopes    []string `json:"scopes"`
	Username  string   `json:"username"`
}

// CreateConfig creates the Config struct from a json file.
func CreateConfig(fileName string) *Config {
	jsonFile, err := os.Open(fileName)
	if err != nil {
		log.Fatal(err)
	}
	defer jsonFile.Close()
	data, err := ioutil.ReadAll(jsonFile)
	if err != nil {
		log.Fatal(err)
	}

	config := &Config{}
	err = json.Unmarshal(data, config)
	if err != nil {
		log.Fatal(err)
	}
	return config
}
```

config.json (anonymized):
```json
{
    "authority": "https://login.microsoftonline.com/<TENANTID>",
    "client_id": "...",
    "scopes": [
        "user.read"
    ],
    "username": "..."
}
```

**Additional context**
I'm using version 1.1.7 of the library because of #380 (my last comment)

**Comment by lestrrat** [comment]
https://github.com/lestrrat-go/jwx/issues/380#issuecomment-860754799

Instead of answering in multiple places, I'm just going to answer here.
I believe this API is okay, because

1. While `WithKeySet` forces you to use a key with the `alg` header, `WithVerify` (...a misnomer, but anyway) allows you to pass whatever key you want as long as you pass the algorithm name with it.
2. If you must use `WithKeySet`, you can always fetch the key, and modify it before passing to `WithKeySet`

The rationale for requiring `alg` is security as you mentioned. But the important distinction from the library's PoV is that we don't provide automatic ways to shoot yourself in the foot: i.e. we're not going to take a key set and automatically assume that both the JWS payload and JWK are well-intentioned players who will not sneak `alg: "none"` or open up the possibility of cheating against potentially weak keys. So `WithKeySet`, which _automatically_ selects a key to use, is not going to assume that checking `alg` can be skipped.

On the other hand, `WithVerify` which takes a single key, is something that the user manually specifies, and therefore we allow the users to shoot themselves in the foot. If you want to fabricate the `alg` value or whatever, you are free to do as you wish.


> Microsoft does not include alg in the key's header. See here:
https://login.microsoftonline.com/common/discovery/v2.0/keys
>
> And it has all the rights to do so, as this key is optional as per RFC:
https://datatracker.ietf.org/doc/html/rfc7517#section-4.4

On this note I think that a) The **RFC is flawed** for allowing this sort of implementation specific loophole/discrepancy, and b) Microsoft should know better to include `alg` -- I mean, it's not going to cause a problem for anybody if the `alg` header is present, but causes problems like this if they don't.

---

So, what can you do on your end? 

First, let me state that I haven't worked with Azure much, so there may be limitation and workarounds that I'm not aware of. Following comes from many assumptions.

So, assuming that Microsoft will not be adding `alg` anytime soon, and that while the keys may be rotated frequently but that their algorithm will stay the same, I would opt to find out the the algorithm used by each of their keys, and hardcode them somewhere  either:

```go
for iter := set.Iterate(ctx); iter.Next(ctx); {
  pair := iter.Pair()
  key := pair.Value.(jwk.Key)
  ...
  tokn, err := jwt.Parse(rawdata, jwt.WithVerify(jwa.RS256, key)
  if err == nil {
     break // do whatever with token
  }
}
```

or

```go
set, _ := af.Fetch(microsoftJwksURL)
// assume you know what algs they were using for each key
key1, _ := set.Get(0)
key1.Set(jwk.AlgorithmKey, jwa.RS256)

key2, _ := set.Get(1)
key2.Set(jwk.AlgorithmKey, jws.HS256)

...
jwt.Parse(rawdata, jwt.WithKeySet(set))
```

**Comment by konradmalik** [comment]
@lestrrat, thank you for such a complete explanation. That clears up the decision from the mentioned PR and allows me to use the latest version of this library with Azure.
However, I still cannot verify tokens from azure due to the mentioned error:
```
failed to parse payload: failed to verify jws signature: failed to verify the message: crypto/rsa: verification error
```

Any idea where to look? What could be wrong? As I said, the token is 100% valid as it allows to reach the Microsoft Graph endpoint using the code provided above without any issues.

The same error appears when I force the proper algorithm on the keys (RS256) or any other wrong.

Keys are loaded properly (I guess). See below:
```
map[e:[1 0 1] issuer:https://login.microsoftonline.com/96327a40-4341-4340-bc1c-1f49cb8f58b3/v2.0 kid:nOo3ZDrODXEK1jKWhXslHR_KXEg kty:RSA n:[161 162 203 79 216 100 113 40 246 180 103 217 178 54 238 237 124 245 42 187 52 168 72 156 92 249 132 176 146 142 5 1 218 185 159 228 68 205 71 116 73 32 56 150 212 206 117 44 164 78 177 58 227 146 94 57 115 18 186 129 197 112 46 225 32 175 114 245 78 9 129 182 191 209 183 250 226 43 44 94 93 178 208 244 177 129 193 163 85 104 154 210 177 204 95 116 86 201 107 146 126 171 53 8 20 24 133 160 190 228 19 198 109 110 70 73 88 20 241 28 76 132 112 208 185 216 146 124 201 211 161 130 246 88 202 71 173 123 192 0 85 173 147 100 6 224 224 64 168 247 238 167 50 198 134 177 32 103 113 17 201 149 21 149 201 202 188 67 129 213 225 246 140 17 153 153 250 44 130 101 26 54 125 155 159 94 153 251 176 156 25 236 33 1 187 14 42 201 193 239 82 94 178 215 157 236 81 65 106 178 204 39 100 97 163 170 146 146 171 178 59 177 82 61 190 48 117 247 212 90 172 118 146 73 176 224 47 97 113 142 96 18 137 70 56 70 172 103 78 53 90 217 117 71 83 5] use:sig x5c:{certs:[0xc00038ac00]} x5t:nOo3ZDrODXEK1jKWhXslHR_KXEg]map[alg:RS256 e:[1 0 1] issuer:https://login.microsoftonline.com/96327a40-4341-4340-bc1c-1f49cb8f58b3/v2.0 kid:nOo3ZDrODXEK1jKWhXslHR_KXEg kty:RSA n:[161 162 203 79 216 100 113 40 246 180 103 217 178 54 238 237 124 245 42 187 52 168 72 156 92 249 132 176 146 142 5 1 218 185 159 228 68 205 71 116 73 32 56 150 212 206 117 44 164 78 177 58 227 146 94 57 115 18 186 129 197 112 46 225 32 175 114 245 78 9 129 182 191 209 183 250 226 43 44 94 93 178 208 244 177 129 193 163 85 104 154 210 177 204 95 116 86 201 107 146 126 171 53 8 20 24 133 160 190 228 19 198 109 110 70 73 88 20 241 28 76 132 112 208 185 216 146 124 201 211 161 130 246 88 202 71 173 123 192 0 85 173 147 100 6 224 224 64 168 247 238 167 50 198 134 177 32 103 113 17 201 149 21 149 201 202 188 67 129 213 225 246 140 17 153 153 250 44 130 101 26 54 125 155 159 94 153 251 176 156 25 236 33 1 187 14 42 201 193 239 82 94 178 215 157 236 81 65 106 178 204 39 100 97 163 170 146 146 171 178 59 177 82 61 190 48 117 247 212 90 172 118 146 73 176 224 47 97 113 142 96 18 137 70 56 70 172 103 78 53 90 217 117 71 83 5] use:sig x5c:{certs:[0xc00038ac00]} x5t:nOo3ZDrODXEK1jKWhXslHR_KXEg]map[alg:RS256 e:[1 0 1] issuer:https://login.microsoftonline.com/96327a40-4341-4340-bc1c-1f49cb8f58b3/v2.0 kid:nOo3ZDrODXEK1jKWhXslHR_KXEg kty:RSA n:[161 162 203 79 216 100 113 40 246 180 103 217 178 54 238 237 124 245 42 187 52 168 72 156 92 249 132 176 146 142 5 1 218 185 159 228 68 205 71 116 73 32 56 150 212 206 117 44 164 78 177 58 227 146 94 57 115 18 186 129 197 112 46 225 32 175 114 245 78 9 129 182 191 209 183 250 226 43 44 94 93 178 208 244 177 129 193 163 85 104 154 210 177 204 95 116 86 201 107 146 126 171 53 8 20 24 133 160 190 228 19 198 109 110 70 73 88 20 241 28 76 132 112 208 185 216 146 124 201 211 161 130 246 88 202 71 173 123 192 0 85 173 147 100 6 224 224 64 168 247 238 167 50 198 134 177 32 103 113 17 201 149 21 149 201 202 188 67 129 213 225 246 140 17 153 153 250 44 130 101 26 54 125 155 159 94 153 251 176 156 25 236 33 1 187 14 42 201 193 239 82 94 178 215 157 236 81 65 106 178 204 39 100 97 163 170 146 146 171 178 59 177 82 61 190 48 117 247 212 90 172 118 146 73 176 224 47 97 113 142 96 18 137 70 56 70 172 103 78 53 90 217 117 71 83 5] use:sig x5c:{certs:[0xc00038ac00]} x5t:nOo3ZDrODXEK1jKWhXslHR_KXEg]map[alg:RS256 e:[1 0 1] issuer:https://login.microsoftonline.com/96327a40-4341-4340-bc1c-1f49cb8f58b3/v2.0 kid:nOo3ZDrODXEK1jKWhXslHR_KXEg kty:RSA n:[161 162 203 79 216 100 113 40 246 180 103 217 178 54 238 237 124 245 42 187 52 168 72 156 92 249 132 176 146 142 5 1 218 185 159 228 68 205 71 116 73 32 56 150 212 206 117 44 164 78 177 58 227 146 94 57 115 18 186 129 197 112 46 225 32 175 114 245 78 9 129 182 191 209 183 250 226 43 44 94 93 178 208 244 177 129 193 163 85 104 154 210 177 204 95 116 86 201 107 146 126 171 53 8 20 24 133 160 190 228 19 198 109 110 70 73 88 20 241 28 76 132 112 208 185 216 146 124 201 211 161 130 246 88 202 71 173 123 192 0 85 173 147 100 6 224 224 64 168 247 238 167 50 198 134 177 32 103 113 17 201 149 21 149 201 202 188 67 129 213 225 246 140 17 153 153 250 44 130 101 26 54 125 155 159 94 153 251 176 156 25 236 33 1 187 14 42 201 193 239 82 94 178 215 157 236 81 65 106 178 204 39 100 97 163 170 146 146 171 178 59 177 82 61 190 48 117 247 212 90 172 118 146 73 176 224 47 97 113 142 96 18 137 70 56 70 172 103 78 53 90 217 117 71 83 5] use:sig x5c:{certs:[0xc00038ac00]} x5t:nOo3ZDrODXEK1jKWhXslHR_KXEg]
```

**Comment by lestrrat** [comment]
Can you please provide a minimal code to reproduce the error? i.e., I assume you tweaked your previous code after my comment. Can you please post that version?

**Comment by konradmalik** [comment]
Sure, only the function `validateToken` from `main.go` was changed:

```go

func validateToken(tokenString string) bool {
	log.Println("Validating token")

	keyset, err := getJWTkeyset()
	if err != nil {
		panic(err)
	}

	for i := 0; i < keyset.Len(); i++ {
		key, _ := keyset.Get(0)
		m, _ := key.AsMap(context.TODO())
		fmt.Printf("%+v", m)
		// assume you know what algs they were using for each key
		key.Set(jwk.AlgorithmKey, jwa.RS256)
	}

	// Actual verification:
	// FINALLY. This is how you Parse and verify the payload.
	// Key IDs are automatically matched.
	// There was a lot of code above, but as a consumer, below is really all you need
	// to write in your code
	token, err := jwt.ParseString(
		tokenString,
		// Tell the parser that you want to use this keyset
		// this will validate signature
		// TODO this does not work! why?
		jwt.WithKeySet(keyset),
		// validate expiration etc.
		jwt.WithValidate(true),
	)
	_ = keyset
	if err != nil {
		fmt.Printf("failed to parse payload: %s\n", err)
		return false
	}
	fmt.Printf("parsed token %v\n", token)
	tid, _ := token.Get("tid")
	fmt.Printf("parsed tid %v\n", tid)
	return true
}
```

**Comment by lestrrat** [comment]
Hmm, bummer. Maybe we "fixed" `WithVerify` where we shouldn't have. Will take a look

**Comment by lestrrat** [comment]
Sorry, can't reproduce. Both `WithVerify` and `WithKeySet` seem to be working as intended.
I don't have an Azure account so I cannot create an Azure-signed token, so I replicated what I think should be happening.

```diff
diff --git a/jwt/jwt_test.go b/jwt/jwt_test.go
index 3241622..83d16af 100644
--- a/jwt/jwt_test.go
+++ b/jwt/jwt_test.go
@@ -914,3 +914,67 @@ func TestJWTParseWithTypedClaim(t *testing.T) {
                })
        }
 }
+
+func TestGH395(t *testing.T) {
+       // Original issue talked about tokens in Microsoft Azure, but in order to "replicate"
+       // what they were doing here we would need an actual JWT signed using their keys...
+       // That is not feasible ATM, so we are just going to sign a random token using our
+       // key, and create a jwk.Set which contains a key without the `alg` field.
+
+       privkey, err := jwxtest.GenerateRsaJwk()
+       if !assert.NoError(t, err, `jwxtest.GenerateRsaJwk() should succeed`) {
+               return
+       }
+       // This guy needs a Key ID (BUT NOT alg)
+       privkey.Set(jwk.KeyIDKey, `my-awesome-key`)
+
+       pubkey, err := jwk.PublicKeyOf(privkey)
+       if !assert.NoError(t, err, `jwk.PublicKeyOf() should succeed`) {
+               return
+       }
+
+       // Make sure we have a key that DOES NOT have the alg field
+       if !assert.Empty(t, pubkey.Algorithm(), `pubkey.Algorithm() should be empty`) {
+               return
+       }
+
+       token := jwt.New()
+       token.Set(jwt.IssuerKey, `https://github.com/lestrrat-go/jwx`)
+
+       signed, err := jwt.Sign(token, jwa.RS256, privkey)
+       if !assert.NoError(t, err, `jwt.Sign() should succeed`) {
+               return
+       }
+
+       t.Run("Using WithVerify()", func(t *testing.T) {
+               // Now make sure we can parse correctly using the public key
+               parsed, err := jwt.Parse(signed, jwt.WithVerify(jwa.RS256, pubkey), jwt.WithValidate(true))
+               if !assert.NoError(t, err, `jwt.Parse() should succeed`) {
+                       return
+               }
+
+               if !assert.True(t, jwt.Equal(parsed, token), "tokens should match") {
+                       return
+               }
+       })
+       t.Run("Using WithKeySet()", func(t *testing.T) {
+               // This time use a key set that contains the public key, and the public key contains
+               // the algorithm used
+               cloned, err := pubkey.Clone()
+               if !assert.NoError(t, err, `pubkey.Clone() should succeed`) {
+                       return
+               }
+               cloned.Set(jwk.AlgorithmKey, jwa.RS256)
+
+               set := jwk.NewSet()
+               set.Add(cloned)
+               parsed, err := jwt.Parse(signed, jwt.WithKeySet(set), jwt.WithValidate(true))
+               if !assert.NoError(t, err, `jwt.Parse() should succeed`) {
+                       return
+               }
+
+               if !assert.True(t, jwt.Equal(parsed, token), "tokens should match") {
+                       return
+               }
+       })
+}

**Comment by konradmalik** [comment]
Well, nevertheless, thank you. I'm connected with this topic for a future couple of days, so maybe I'll come up with something useful and post it here.

**Comment by lestrrat** [comment]
Well, FWIW, after revisiting your error message in https://github.com/lestrrat-go/jwx/issues/395#issuecomment-861453772, I *think* it's finding your key (the error is from crypto/rsa, which means it _is_ passing a key to verify with) so It's just failing to verify, which just means the keys are wrong or the token is wrong.

**Comment by konradmalik** [comment]
To be sure - `parseString` accepts base64-encoded token as in, for example, `Authorization: Bearer xxx` header, right?

**Comment by lestrrat** [comment]
No :)  it expects JWS compact or JWS JSON format message.

Incidentally, you made me realized that jwt.ParseRequest has a bug.

**Comment by konradmalik** [comment]
haaa... that clears a bit. Does the bug you mentioned concerns also ParseHeader? It seems that all ParseHeader does, it strips Bearer and passes the token to ParseString. However, you just told me that ParseString wouldn't handle base64 encoded token, but that is exactly the case here, right?

**Comment by lestrrat** [comment]
you are not using ParseHeader, so what I weote doesn't matter in your case. The "bug" is more of a ambiguity in the spec. Some implementations MAY actually encode the jws in base64, and there should be a way to handlw it.

**Comment by lestrrat** [comment]
At this point I would probably print out the jws message you want to verify, and create an isolated test case using the actual signed value. As it is I can't do mich because I don't know what you are parsing.

**Comment by konradmalik** [comment]
Reg. ParseHeader - in this example I did not use it, true, but I did in other tests in the meantime.

However the puzzle is solved already, and of course nothing is wrong with the library ;)

TLDR; Azure provided 2 different kinds of tokens. One is AccessToken (the one I used), the other one is IDToken. The former one is not exactly (although VERY similar) an JWT token. Signature is different, and requires special handling during verification. The latter one is a standard JWT token and can be successfully verified using standard methods.
So, in the code above, the solution was to change `authResult.AccessToken` to `authResult.IDToken.RawToken` and all works flawlessly. Note that the above-mentioned workaround for the key algorithms is required, nothing changes in that regard.

More info here:
- https://stackoverflow.com/questions/60778634/failing-signature-validation-of-jwt-tokens-from-azure-ad
- https://docs.microsoft.com/en-us/azure/active-directory-b2c/tokens-overview

**Comment by lestrrat** [comment]
@konradmalik I had an epiphany yesterday, and wrote something up that could have helped you on this

https://github.com/lestrrat-go/jwx/tree/infer-algorithm

This is still WIP, but this would automatically pick a key to use with JWKS without the alg field present. I'm still pondering whether if this is right, but if you interested I'm all ears for feedback

**Comment by lestrrat** [comment]
For anyone who came here late to the game, this has been merged as #470

**Comment by konradmalik** [comment]
Hey @lestrrat, sorry I’ve been unresponsive, but I had to settle other things. I’ll try this out and leave some comments here today/tomorrow.

**Comment by konradmalik** [comment]
Soo, it works. Allowed me to save 1 function that I used to extract the key from the header and some logic here and there.

Overall looks good, however, maybe the user should be able to pass a list of "expected" algorithms?
Not sure about that, just wondering out loud, however the fact is that:
https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration - specifies `id_token_signing_alg_values_supported":["RS256"]` which is just one key (can be multiple sometime in the future?).
So, someone could read this configuration once in a while, and while checking the keys, check only of those provided in the "supported array".
What do you think?

**Comment by lestrrat** [comment]
Hmm, this is just my initial reaction, but I don't immediately see a good API for that.

Suppose you have a JWKS with keys for [ RS256 ], [ RS512, RS384, RS256 ], and [ PS512 ]
How do you specify the set of algorithms as parameters to `jwt.Parse`?

Obviously, if you _know_ before hand that you are only ever going to use one of the keys, you can do something like

```go
jwt.Parse(..., jwt.WithAlgorithms(jwa.RS256), jwt.WithKeySet(ks))
```

But what if you only know that it's going to be one of the three, then you would be passing a list like

```go
jwt.Parse(..., jwt.WithAlgorithms(jwa.RS256, jwa.RS384, jwa.RS512, jwa.PS256), jwt.WithKeySet(ks))
```

...Does this look like an improvement? What if the key types were mixed. It will be even worse:

```go
jwt.Parse(..., jwt.WithAlgorithms(jwa.RS256, jwa.EdDSA, jwa.RS512, jwa.HS256, jwa.RS256), jwt.WithKeySet(ks))
```

I mean, what is this? :)

It would be much more efficient if you could specify the algorithms per key, but this, I really can't think of any good API.

At this point I would then just say, do this:

```go
for i := 0; i < ks.Len(); i++ {
  key := ks.Get(i)
  for _, alg := range knownAlgorithmsForType[key] { // your "optimized" list of algorithms
      jwt.Parse(..., jwt.WithVerify(alg, key))
  }
}
````

So at the moment I can't think of a good, generic API that would allow you to do what you want, if I understand the problem correctly.

**Comment by konradmalik** [comment]
@lestrrat sure, all makes sense. I still think that given that we have some knowledge about the possible keys, we should be able to "inform" the library about it somehow. It just seems logical and should increase security in this fairly insecure situation. But I understand and agree with your point, and I have to say I also don't know how to do it cleanly.
